### PR TITLE
BIGTOP-3967: add support for spark

### PR DESCRIPTION
**Pull request by MacChen01**
### Description of PR
Add support for spark component, include compile command and problem.
1. install R by dnf in openEuler
2. add pandoc by wget and install in openEuler

### How was this patch tested?
build command
docker run --rm -v pwd:/home/bigtop --workdir /home/bigtop bigtop/slaves:trunk-openeuler-22.03-aarch64 bash -c '. /etc/profile.d/bigtop.sh; ./gradlew spark-pkg'

smoke test command
./docker-hadoop.sh -C config_openeuler-22.03.yaml -c 3 -s

### For code changes:

- [ ] Does the title or this PR starts with the corresponding JIRA issue id (BIGTOP-3967)
- [ ] Make sure that newly added files do not have any licensing issues. When in doubt refer to https://www.apache.org/licenses/

### CODE REVIEW

**Comment by MacChen01** [comment]
**Update the spec file(commit: https://github.com/apache/bigtop/pull/1148/commits/a0a68f101bd48edfe1ef0376c43c034f328107a4), becasue we met the error in rockyLinux**:
image

**We found the _%elif_ need the version of RPM is 4.15.0+, but rockyLinux's version is 4.14.3**
image
image

**Comment by MacChen01** [comment]
**Spark smoke test successful log on openEuler**
image

The detail log:
https://gitee.com/macchen1/bigtop/blob/master/smoke-test-log/openeuler-smoke-log/spark-openeuler.log

**Comment by MacChen01** [comment]
Spark smoke test successful log on RocklyLinux
image

The detail log：
https://gitee.com/macchen1/bigtop/blob/master/smoke-test-log/rocklylinux-8-smoke-log/spark-rocklylinux-8.log

@guyuqi

File: bigtop-deploy/puppet/modules/spark/manifests/init.pp [hunk 1]
```diff
@@ -202,6 +202,20 @@
...
+      'x86_64' : { $arch= "x86_64"}
+    }
+
+    if ($operatingsystem == 'openEuler'){
+       exec {"install_R":
+         command => "/usr/bin/dnf install 'dnf-command(config-manager)' -y && /usr/bin/dnf config-manager --add-repo https://repo.oepkgs.net/openeuler/rpm/openEuler-22.03-LTS/extras/$arch && /usr/bin/dnf clean all && /usr/bin/dnf makecache && /usr/bin/dnf install -y R",
+         before  => [Package["spark-sparkr"]],
+         timeout => 6000
+       }
+    }
```
Comment (guyuqi) [inline bigtop-deploy/puppet/modules/spark/manifests/init.pp:217]: Please move `R` installation into `bigtop_toolchain`.

File: bigtop-packages/src/common/spark/do-component-build [hunk 2]
```diff
@@ -30,6 +30,22 @@ BUILD_OPTS="-Divy.home=${HOME}/.ivy2 -Dsbt.ivy.home=${HOME}/.ivy2 -Duser.home=${
...
+    wget https://github.com/jgm/pandoc/releases/download/2.19.2/pandoc-2.19.2-linux-arm64.tar.gz
+    tar -xf pandoc-2.19.2-linux-arm64.tar.gz
+    cp -r pandoc-2.19.2/bin/pandoc /usr/bin/
+  else
+    wget https://github.com/jgm/pandoc/releases/download/2.19.2/pandoc-2.19.2-linux-amd64.tar.gz
+    tar -xf pandoc-2.19.2-linux-amd64.tar.gz
+    cp -r pandoc-2.19.2/bin/pandoc /usr/bin/
+  fi
+fi
+
```
Comment (guyuqi) [inline bigtop-packages/src/common/spark/do-component-build:48]: Please also move `pandoc` into `bigtop_toolchain`.
https://github.com/apache/bigtop/blob/master/bigtop_toolchain/manifests/renv.pp#L25

Comment (guyuqi) [reply bigtop-packages/src/common/spark/do-component-build:48]: We install R and Pandoc in Bigtop_toolchain.
Please do not install the dependency packages when building components.
@MacChen01

**Comment by Hann0321** [comment]
@guyuqi

**Comment by Hann0321** [comment]
@guyuqi 
When I test the bigtop smoke test of the Ubuntu 22.04 system, the apt repository cannot be found.
image
I build the local repository source command:  ./gradlew apt
Can you provide a way to test the smoke test of bigtop on Ubuntu 22.04?

**Comment by guyuqi** [comment]
> @guyuqi When I test the bigtop smoke test of the Ubuntu 22.04 system, the apt repository cannot be found. image I build the local repository source command: ./gradlew apt Can you provide a way to test the smoke test of bigtop on Ubuntu 22.04?


In Ubuntu-22.04 yaml: 
`https://github.com/apache/bigtop/blob/master/provisioner/docker/config_ubuntu-22.04.yaml`
You should enable your local repo and set the components properly like:
```
components: [hdfs, yarn, mapreduce, hbase]
enable_local_repo: true
smoke_test_components: [hbase]
```

To run the smoke test:
`./docker-hadoop.sh -C config_ubuntu-22.04.yaml -c 3 -s`

**Comment by Hann0321** [comment]
> 
@guyuqi 
My operation is the same as you said, but it's wrong.
image

**Comment by Hann0321** [comment]
> > @guyuqi When I test the bigtop smoke test of the Ubuntu 22.04 system, the apt repository cannot be found. image I build the local repository source command: ./gradlew apt Can you provide a way to test the smoke test of bigtop on Ubuntu 22.04?
> 
> In Ubuntu-22.04 yaml: `https://github.com/apache/bigtop/blob/master/provisioner/docker/config_ubuntu-22.04.yaml` You should enable your local repo and set the components properly like:
> 
> ```
> components: [hdfs, yarn, mapreduce, hbase]
> enable_local_repo: true
> smoke_test_components: [hbase]
> ```
> 
> To run the smoke test: `./docker-hadoop.sh -C config_ubuntu-22.04.yaml -c 3 -s`

@guyuqi 
My operation is the same as you said, but it's wrong.
image

**Comment by guyuqi** [comment]
I just successfully verified `zookeeper` on my local ubuntu-22 environment:
```
components: [zookeeper]
enable_local_repo: true
smoke_test_components: [zookeeper]
```

The yaml should be:
```
docker:
        memory_limit: "4g"
        image:  "bigtop/puppet:trunk-ubuntu-22.04"
        repo: "http://repos.bigtop.apache.org/releases/3.2.0/ubuntu/22.04/$(ARCH)"
distro: debian
components: [zookeeper]
enable_local_repo: true
smoke_test_components: [zookeeper]
```

You should not modify the field: `repo`.
The field `enable_local_repo: true` has enabled your local repo.

If we enable local repo, we should also build the prerequisite components : `bigtop-utils`, `bigtop-jsvc` and `bigtop-groovy` and don't forget to setup apt repo at last.
`./gradlew bigtop-utils-pkg bigtop-groovy-pkg bigtop-jsvc-pkg zookeeper-pkg apt`

**Comment by Hann0321** [comment]
> ./gradlew bigtop-utils-pkg bigtop-groovy-pkg bigtop-jsvc-pkg

I have built all prerequisite components and apt, but the error still occurs.
image
the error:
image

**Comment by guyuqi** [comment]
The related field `repo` should not be changed to your local repo file:///bigtop-home/output

> > ./gradlew bigtop-utils-pkg bigtop-groovy-pkg bigtop-jsvc-pkg
> 
> I have built all prerequisite components and apt, but the error still occurs. image the error: image

May I ask how did you build these components? 
1. Please build the components based in docker images `bigtop/slaves:3.2.0-ubuntu-22.04-aarch64`.
2. Please build the components under the user: `jenkins`.

File: bigtop_toolchain/manifests/renv.pp [hunk 3]
```diff
@@ -49,22 +49,42 @@
...
-      ensure => installed,
-      before => [Exec["install_r_packages"]]
+  if ($operatingsystem == 'openEuler'){
+    if ($architecture == "aarch64") {
+        $url = "https://github.com/jgm/pandoc/releases/download/2.19.2/pandoc-2.19.2-linux-arm64.tar.gz"
+        $pandoctar = "pandoc-2.19.2-linux-arm64.tar.gz"
+    } else{
+        $url = "https://github.com/jgm/pandoc/releases/download/2.19.2/pandoc-2.19.2-linux-amd64.tar.gz"
+        $pandoctar = "pandoc-2.19.2-linux-amd64.tar.gz"
+    }
```
Comment (guyuqi) [inline bigtop_toolchain/manifests/renv.pp:67]: Does openEuler offically support PowerPC?
If not, please add comments here and Bigtop would not offer the openEuler packages for ppc64le in the future.
(We would add this notification in the next release annoucement).

Comment (MacChen01) [reply bigtop_toolchain/manifests/renv.pp:67]: openEuler currently does not support PowerPC, i will add comments later.

File: bigtop-deploy/puppet/modules/spark/manifests/init.pp (hunk 1 above)
Comment (MacChen01) [reply bigtop-deploy/puppet/modules/spark/manifests/init.pp]: R installation moved into bigtop_toolchain.
commits: https://github.com/apache/bigtop/pull/1148/commits/6d5fa5ad7ff65d03b9836f074b2129e5fb5a6c5d

File: bigtop-packages/src/common/spark/do-component-build (hunk 2 above)
Comment (MacChen01) [reply bigtop-packages/src/common/spark/do-component-build]: Pandoc installation moved into bigtop_toolchain. @guyuqi 
commits: https://github.com/apache/bigtop/pull/1148/commits/6d5fa5ad7ff65d03b9836f074b2129e5fb5a6c5d

**Comment by MacChen01** [comment]
We moved R and Pandoc into bigtop_toolchain, compile and smoke test successfully.
**The openEuler successful log:**
image
**The detail log:**
https://gitee.com/macchen1/bigtop/blob/master/smoke-test-log/openeuler-smoke-log/spark-openeuler-change-to-bigtop_toolchain.log
**The rocklylinux successful log:**
image
**The detail log:**
https://gitee.com/macchen1/bigtop/blob/master/smoke-test-log/rocklylinux-8-smoke-log/spark-change-bigop_tooltain-rockly-8.log

@guyuqi

**Comment by MacChen01** [comment]
@guyuqi Thanks for your help and advice.
1. Moved R and pandoc installation into bigtop_toolchain, smoke test successfully: https://github.com/apache/bigtop/pull/1148#issuecomment-1665288175
2. Add comments for openEuler currently does not support PowerPC. (commit:https://github.com/apache/bigtop/commit/d2e8e4ecfc96b2208cc97381c2978c56a53ded72)
3. The discussion about compile/smokeTest is ok on **_ubuntu-22.04_**, ambari/kafka/solr/zk smoke test successfully, you can get the detail log: https://gitee.com/macchen1/bigtop/tree/master/smoke-test-log/ubuntu-22-smoke-log
 and we will push the other smoke test log on it.

**Comment by guyuqi** [comment]
> @guyuqi Thanks for your help and advice.
> 
> 1. Moved R and pandoc installation into bigtop_toolchain, smoke test successfully: [BIGTOP-3967: add support for spark #1148 (comment)](https://github.com/apache/bigtop/pull/1148#issuecomment-1665288175)
> 2. Add comments for openEuler currently does not support PowerPC. (commit:[d2e8e4e](https://github.com/apache/bigtop/commit/d2e8e4ecfc96b2208cc97381c2978c56a53ded72))
> 3. The discussion about compile/smokeTest is ok on **_ubuntu-22.04_**, ambari/kafka/solr/zk smoke test successfully, you can get the detail log: https://gitee.com/macchen1/bigtop/tree/master/smoke-test-log/ubuntu-22-smoke-log
>    and we will push the other smoke test log on it.


I rebuilt openEuler docker image and tried to build Hadoop and Spark with this image.
It failed to build Hadoop:
```
[INFO] [2/4] Fetching packages...
[INFO] error triple-beam@1.4.1: The engine "node" is incompatible with this module. Expected version ">= 14.0.0". Got "12.22.1"
[INFO] error Found incompatible module.
[INFO] error Found incompatible module.info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Hadoop Main 3.3.5:
...
.
.
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  27:38 min
[INFO] Finished at: 2023-08-15T06:00:24Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.11.2:yarn (yarn install) on project hadoop-yarn-applications-catalog-webapp: Failed to run task: 'yarn ' failed. org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :hadoop-yarn-applications-catalog-webapp
```
@MacChen01

**Comment by MacChen01** [comment]
> > @guyuqi Thanks for your help and advice.
> > 
> > 1. Moved R and pandoc installation into bigtop_toolchain, smoke test successfully: [BIGTOP-3967: add support for spark #1148 (comment)](https://github.com/apache/bigtop/pull/1148#issuecomment-1665288175)
> > 2. Add comments for openEuler currently does not support PowerPC. (commit:[d2e8e4e](https://github.com/apache/bigtop/commit/d2e8e4ecfc96b2208cc97381c2978c56a53ded72))
> > 3. The discussion about compile/smokeTest is ok on **_ubuntu-22.04_**, ambari/kafka/solr/zk smoke test successfully, you can get the detail log: https://gitee.com/macchen1/bigtop/tree/master/smoke-test-log/ubuntu-22-smoke-log
> >    and we will push the other smoke test log on it.
> 
> I rebuilt openEuler docker image and tried to build Hadoop and Spark with this image. It failed to build Hadoop:
> 
> ```
> [INFO] [2/4] Fetching packages...
> [INFO] error triple-beam@1.4.1: The engine "node" is incompatible with this module. Expected version ">= 14.0.0". Got "12.22.1"
> [INFO] error Found incompatible module.
> [INFO] error Found incompatible module.info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Summary for Apache Hadoop Main 3.3.5:
> ...
> .
> .
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  27:38 min
> [INFO] Finished at: 2023-08-15T06:00:24Z
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.11.2:yarn (yarn install) on project hadoop-yarn-applications-catalog-webapp: Failed to run task: 'yarn ' failed. org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
> [ERROR]
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR]
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
> [ERROR]
> [ERROR] After correcting the problems, you can resume the build with the command
> [ERROR]   mvn <args> -rf :hadoop-yarn-applications-catalog-webapp
> ```
> 
> @MacChen01

This bug?
https://issues.apache.org/jira/browse/BIGTOP-3960
@guyuqi

**Comment by guyuqi** [comment]
> > > @guyuqi Thanks for your help and advice.
> > > 
> > > 1. Moved R and pandoc installation into bigtop_toolchain, smoke test successfully: [BIGTOP-3967: add support for spark #1148 (comment)](https://github.com/apache/bigtop/pull/1148#issuecomment-1665288175)
> > > 2. Add comments for openEuler currently does not support PowerPC. (commit:[d2e8e4e](https://github.com/apache/bigtop/commit/d2e8e4ecfc96b2208cc97381c2978c56a53ded72))
> > > 3. The discussion about compile/smokeTest is ok on **_ubuntu-22.04_**, ambari/kafka/solr/zk smoke test successfully, you can get the detail log: https://gitee.com/macchen1/bigtop/tree/master/smoke-test-log/ubuntu-22-smoke-log
> > >    and we will push the other smoke test log on it.
> > 
> > 
> > I rebuilt openEuler docker image and tried to build Hadoop and Spark with this image. It failed to build Hadoop:
> > ```
> > [INFO] [2/4] Fetching packages...
> > [INFO] error triple-beam@1.4.1: The engine "node" is incompatible with this module. Expected version ">= 14.0.0". Got "12.22.1"
> > [INFO] error Found incompatible module.
> > [INFO] error Found incompatible module.info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> > [INFO] ------------------------------------------------------------------------
> > [INFO] Reactor Summary for Apache Hadoop Main 3.3.5:
> > ...
> > .
> > .
> > [INFO] BUILD FAILURE
> > [INFO] ------------------------------------------------------------------------
> > [INFO] Total time:  27:38 min
> > [INFO] Finished at: 2023-08-15T06:00:24Z
> > [INFO] ------------------------------------------------------------------------
> > [ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.11.2:yarn (yarn install) on project hadoop-yarn-applications-catalog-webapp: Failed to run task: 'yarn ' failed. org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
> > [ERROR]
> > [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> > [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> > [ERROR]
> > [ERROR] For more information about the errors and possible solutions, please read the following articles:
> > [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
> > [ERROR]
> > [ERROR] After correcting the problems, you can resume the build with the command
> > [ERROR]   mvn <args> -rf :hadoop-yarn-applications-catalog-webapp
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @MacChen01
> 
> This bug? https://issues.apache.org/jira/browse/BIGTOP-3960 @guyuqi

oops, I forgot that I have fixed this issue in https://github.com/apache/bigtop/pull/1133.
Let me cherry pick it to openEuler Branch.

**Comment by guyuqi** [comment]
> Spark smoke test successfully using bigtop/puppet, the log is: image
> 
> **The detail log is:** https://gitee.com/macchen1/bigtop/blob/master/smoke-test-log/openeuler-puppet-smoke-log/spark-openeuler.log
> 
> @guyuqi pls help to check the code we push, thanks.


Let me re-build the bigtop_toolchain images and verify the PR #1148 #1147 #1143 together.
Thanks.
@MacChen01

**Comment by MacChen01** [comment]
> > Spark smoke test successfully using bigtop/puppet, the log is: image
> > **The detail log is:** https://gitee.com/macchen1/bigtop/blob/master/smoke-test-log/openeuler-puppet-smoke-log/spark-openeuler.log
> > @guyuqi pls help to check the code we push, thanks.
> 
> Let me re-build the bigtop_toolchain images and verify the PR #1148 #1147 #1143 together. Thanks. @MacChen01

ok, thanks.

**Comment by guyuqi** [comment]
It failed to build Spark based on openEuler toolchain images:
```
...

Error in loadVignetteBuilder(pkgdir, TRUE) :
  vignette builder 'knitr' not found
Execution halted
error: Bad exit status from /var/tmp/rpm-tmp.6j91xO (%build)
...
```

Kengo had fixed the similar issue in https://issues.apache.org/jira/browse/BIGTOP-3540.
It appears `R` was not installed in openEuler properly. 
Please refer to PR #776 .

Thanks. @MacChen01

**Comment by guyuqi** [comment]
I applied the patches from PRs:  #1148, #1147 and #1143 yesterday.
Built toolchain images: `bigtop/puppet:est-openeuler-22.03-aarch64` and `bigtop/slaves:test-openeuler-22.03-aarch64`.

Built Spark in `bigtop/slaves:test-openeuler-22.03-aarch64`.
@uscxiexing

**Comment by guyuqi** [comment]
I also cherry-pick some patches from `master` to branch `openEuler-support` several days ago.
Please keep your local branch updated. Thanks.

**Comment by guyuqi** [comment]
I'd like to build toochain and Spark again in a new node and make double check.
I will update the status here later. 
Thanks.
@uscxiexing

**Comment by uscxiexing** [comment]
> 


> I'd like to build toochain and Spark again in a new node and make double check. I will update the status here later. Thanks. @uscxiexing

Thank you, I'm also re-running this pr and build spark.

**Comment by MacChen01** [comment]
> It failed to build Spark based on openEuler toolchain images:
> 
> ```
> ...
> 
> Error in loadVignetteBuilder(pkgdir, TRUE) :
>   vignette builder 'knitr' not found
> Execution halted
> error: Bad exit status from /var/tmp/rpm-tmp.6j91xO (%build)
> ...
> ```
> 
> Kengo had fixed the similar issue in https://issues.apache.org/jira/browse/BIGTOP-3540. It appears `R` was not installed in openEuler properly. Please refer to PR #776 .
> 
> Thanks. @MacChen01

We didn't modify the code about this, as you see the bugfix has merged into master branch, please help to check it.
And we will check and verfiy again.
@guyuqi

**Comment by uscxiexing** [comment]
@guyuqi 
we checked and verfiy again.Build  spark successful ，and smoke test is also successful.
**compile spark successful log**:
the time zone is UTC .
image

 **smoke test  successful log:**
image

**Comment by MacChen01** [comment]
we check and verfiy successfully.
@guyuqi

**Comment by guyuqi** [comment]
> we check and verfiy successfully. @guyuqi

Sorry, I have been caught up with other things. I will test it ASAP.

**Comment by guyuqi** [comment]
PR #1143 #1147 and #1148 have been merged.
Thanks, @MacChen01 @uscxiexing

**Comment by MacChen01** [comment]
> PR #1143 #1147 and #1148 have been merged. Thanks, @MacChen01 @uscxiexing

Thanks @guyuqi

**Pull request merged**